pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/ReplicatedSubscriptionSnapshotCache.java

```diff
@@ -50,12 +57,60 @@ public synchronized void addNewSnapshot(ReplicatedSubscriptionsSnapshot snapshot
                     snapshot.getSnapshotId());
         }
 
-        snapshots.put(position, snapshot);
+        if (positionToTimestamp.lastEntry() == null) {
+            snapshots.put(position, snapshot);
+            positionToTimestamp.put(position, publishTime);
+            return;
+        }
+
+        final long timeSinceFirstSnapshot = publishTime - positionToTimestamp.firstEntry().getValue();
+        final long timeSinceLastSnapshot = publishTime - positionToTimestamp.lastEntry().getValue();
+        final long timeWindowPerSlot = timeSinceFirstSnapshot / snapshotFrequencyMillis / maxSnapshotToCache;
+        // reset cursor
+        if (position.compareTo(positionToTimestamp.firstKey()) < 0) {
+            positionToTimestamp.clear();
+            snapshots.clear();
+            snapshots.put(position, snapshot);
+            positionToTimestamp.put(position, publishTime);
+            return;
+        } else if (position.compareTo(positionToTimestamp.lastKey()) < 0) {
+            while (position.compareTo(positionToTimestamp.lastKey()) < 0) {
+                positionToTimestamp.pollLastEntry();
+                snapshots.pollLastEntry();
+            }
+            snapshots.put(position, snapshot);
+            positionToTimestamp.put(position, publishTime);
+        }
+
+        if (timeSinceLastSnapshot < snapshotFrequencyMillis || timeSinceLastSnapshot < timeWindowPerSlot) {
+            return;
+        }
+        if (snapshots.size() < maxSnapshotToCache) {
+            snapshots.put(position, snapshot);
+            positionToTimestamp.put(position, publishTime);
+        } else {
+            int medianIndex = maxSnapshotToCache / 2;
+            Position positionToRemove = findPositionByIndex(medianIndex);
+            if (positionToRemove != null) {
+                positionToTimestamp.remove(positionToRemove);
+                snapshots.remove(positionToRemove);
+            }
+            positionToTimestamp.put(position, publishTime);
+            snapshots.put(position, snapshot);
+        }
+    }
```

Comment:
Please add sufficient code comments so that it's possible to read the code and understand the algorithm. It might be worth extracting this logic to a method and add a comment on the method with a longer description.

Reply:
Good idea